packages/gatsby-link/src/index.tsx

```diff
@@ -88,9 +161,9 @@ class GatsbyLink extends React.Component {
     instance.disconnect()
   }
 
-  handleRef(ref) {
-    if (this.props.innerRef && this.props.innerRef.hasOwnProperty(`current`)) {
-      this.props.innerRef.current = ref
+  handleRef: React.Ref<HTMLAnchorElement> = (ref: HTMLAnchorElement) => {
+    if (refIsObject(this.props.innerRef)) {
+      asMutable(this.props.innerRef).current = ref
```

Comment:
I am severely confused by this entire `handlRef` function... Like we shouldn't need to do any of this. We should just be passing `innerRef={this.props.ref}` in the below render. Then move the createIntersectionObserver to a componentDidUpdate or something. 

I also don't see when `this.props.innerRef` would be a function. Or why it is.. idk (maybe it's just late and I'm tired 😆 )

Reply:
Yeah so Ref can be either a function or an object, so I think someone passing GatsbyLink a function ref would do it? Not sure.

Regarding your first point... yeah... the logic does feel convoluted. I was hoping to avoid changing it from what it was initially, so that I don't accidentally change the behavior, but if someone more intimiately familiar with the original intent can vet a refactor, then it might make sense!